admin_pages/events/Events_Admin_Page.core.php

```diff
@@ -2104,35 +2116,167 @@ protected function _delete_event($redirect_after = true)
+        }
+        wp_safe_redirect(
+            EE_Admin_Page::add_query_args_and_nonce(
+                $args,
+                $this->_admin_base_url
+            )
+        );
+    }
+
+    /**
+     * A page for users to preview what exactly will be deleted, and confirm they want to delete it.
+     * @since $VID:$
+     */
+    protected function previewDeletion()
+    {
+        $EVT_IDs = isset($this->_req_data['EVT_IDs']) ? (array) $this->_req_data['EVT_IDs'] : array();
+        $confirm_deletion_args = [
+            'action' => 'confirm_deletion',
+        ];
         foreach ($EVT_IDs as $EVT_ID) {
-            $EVT_ID = absint($EVT_ID);
-            if ($EVT_ID) {
-                $results = $this->_permanently_delete_event($EVT_ID);
-                $success = $results !== false ? $success : false;
-                // remove this event from the list of events with no prices
-                unset($espresso_no_ticket_prices[ $EVT_ID ]);
+            $confirm_deletion_args['EVT_IDs[]'] = (int) $EVT_ID;
+        }
+        $this->_template_args['admin_page_content'] = EEH_Template::display_template(
+            EVENTS_TEMPLATE_PATH . 'event_preview_deletion.template.php',
+            [
+                'form_url' => EE_Admin_Page::add_query_args_and_nonce(
+                    $confirm_deletion_args,
+                    $this->admin_base_url()
+                )
+            ],
+            true
+        );
+        $this->display_admin_page_with_no_sidebar();
+    }
+
+    protected function confirmDeletion()
+    {
+        $event_ids = isset($this->_req_data['EVT_IDs']) ? $this->_req_data['EVT_IDs'] : array();
+        $success = $this->deleteEventsAndDependentData($event_ids);
+        $this->redirect_after_action(
+            $success,
+            esc_html__('Events', 'event_espresso'),
+            esc_html__('deleted', 'event_espresso'),
+            [
+                'action' => 'default'
+            ]
+        );
+    }
+
+    /**
+     * Deletes the events and all dependent data, plus those events' non-global/non-default tickets, prices, and
+     * message template groups.
+     * @since $VID:$
+     * @param $event_ids
+     * @return int
+     * @throws EE_Error
+     * @throws InvalidArgumentException
+     * @throws ReflectionException
+     * @throws InvalidDataTypeException
+     * @throws InvalidInterfaceException
+     */
+    protected function deleteEventsAndDependentData($event_ids)
+    {
+        // Call me an optimist.
+        $success = true;
+        $espresso_no_ticket_prices = get_option('ee_no_ticket_prices', array());
+
+        // Find all the root nodes to delete (this isn't just events, because there's other data, like related tickets,
+        // prices, message templates, etc, whose model definition doesn't make them dependent on events. But,
+        // we have no UI to access them independent of events, so they may as well get deleted too.)
```

Comment:
Right, tickets, prices and message templates are hard-coded because, according to the models, they don't actually depend on events. 
But other models that DO depend on events, they don't need to be hard-coded, and are deleted through the tree traversal code done earlier (see https://github.com/eventespresso/event-espresso-core/issues/349#issuecomment-548505642). Specifically, these things gets deleted:
-Registration
--Answer
--Checkin
--Registration_Payment
--(not Messages because they don't block deletions)
-Datetime
--Checkin
--Datetime_Ticket
-Event_Question_Group
-Term_Relationship
-Event_Message_Template_Group